lib/hooks/trade.tsx

```diff
@@ -0,0 +1,238 @@
+  const { data: itemState } = useTradeItemState(item);
+
+  const maxBaseAmount = useTradeMaxBaseAmount(item);
+
+  if (!itemState) {
+    return new Decimal(0);
+  }
+
+  const {
+    poolAssetBalance,
+    baseWeight,
+    poolBaseBalance,
+    assetWeight,
+    tradeablePoolAssetBalance,
+    swapFee,
+    traderBaseBalance,
+  } = itemState;
+
+  let maxAmountAsset: Decimal;
+
+  if (item.action === "buy") {
+    maxAmountAsset = calcOutGivenIn(
+      poolBaseBalance,
+      baseWeight,
+      poolAssetBalance,
+      assetWeight,
+      maxBaseAmount,
+      swapFee,
+    );
+  }
+  if (item.action === "sell") {
+    maxAmountAsset = calcInGivenOut(
+      poolAssetBalance,
+      assetWeight,
+      poolBaseBalance,
+      baseWeight,
+      maxBaseAmount,
+      swapFee,
+    );
+  }
+
+  maxAmountAsset =
+    traderBaseBalance == null
+      ? new Decimal(0)
+      : maxAmountAsset.gt(tradeablePoolAssetBalance)
+      ? tradeablePoolAssetBalance
+      : maxAmountAsset;
+
+  return maxAmountAsset;
+};
+
+export const useTradeTransaction = (
+  item: TradeItem,
+  amountAsset?: string,
+): SubmittableExtrinsic<"promise", ISubmittableResult> | undefined => {
+  amountAsset = !amountAsset ? "0" : amountAsset;
+  const amountAssetDecimal = new Decimal(amountAsset).mul(ZTG);
+  const [sdk] = useSdkv2();
+  const { data: itemState } = useTradeItemState(item);
+
+  if (itemState == null || sdk == null) {
+    return undefined;
+  }
+
+  const {
+    pool,
+    poolAssetBalance,
+    baseWeight,
+    poolBaseBalance,
+    assetWeight,
+    swapFee,
+    slippage,
+  } = itemState;
+
+  let transaction:
+    | SubmittableExtrinsic<"promise", ISubmittableResult>
+    | undefined;
+
+  if (isRpcSdk(sdk)) {
+    try {
+      if (item.action == "buy") {
+        const maxAmountIn = calcInGivenOut(
+          poolBaseBalance,
+          baseWeight,
+          poolAssetBalance,
+          assetWeight,
+          amountAssetDecimal,
+          swapFee,
+        ).mul(new Decimal(slippage / 100 + 1));
+
+        if (!maxAmountIn.isNaN()) {
+          transaction = sdk.api.tx.swaps.swapExactAmountOut(
+            pool.poolId,
+            { Ztg: null },
+            maxAmountIn.toFixed(0),
+            item.assetId,
+            amountAssetDecimal.toFixed(0),
+            null,
+          );
+        }
+      } else {
+        const minAmountOut = calcOutGivenIn(
+          poolAssetBalance,
+          assetWeight,
+          poolBaseBalance,
+          baseWeight,
+          amountAssetDecimal,
+          swapFee,
+        ).mul(new Decimal(1 - slippage / 100));
+
+        if (!minAmountOut.isNaN()) {
+          transaction = sdk.api.tx.swaps.swapExactAmountIn(
+            pool.poolId,
+            item.assetId,
+            amountAssetDecimal.toFixed(0),
+            { Ztg: null },
+            minAmountOut.toFixed(0),
+            null,
+          );
+        }
+      }
+    } catch (error) {
+      console.log(error);
```

Comment:
I think it's best to throw an error here or not catch at all, otherwise it won't get picked up by react-query. https://tanstack.com/query/v4/docs/react/guides/query-functions#handling-and-throwing-errors